autofaiss/external/distributed.py

```diff
@@ -0,0 +1,349 @@
+"""
+Building the index with pyspark.
+"""
+import math
+import multiprocessing
+import os
+import shutil
+from itertools import repeat
+from multiprocessing.pool import ThreadPool
+from tempfile import NamedTemporaryFile
+from typing import List, Optional, Union, Iterator, Tuple, Callable, Any
+
+import faiss
+import fsspec
+import numpy as np
+import pandas as pd
+from fsspec import get_filesystem_class
+from tqdm import tqdm
+
+from autofaiss.readers.embeddings_iterators import get_file_list, get_matrix_reader
+from autofaiss.utils.decorators import Timeit
+
+
+def _yield_embeddings_batch(
+    embeddings_paths: List[str],
+    chunk_sizes: List[int],
+    start: int,
+    end: int,
+    embedding_column_name: str,
+    file_format: str,
+    id_columns: Optional[List[str]] = None,
+):
+    """Lazy load a batch of embeddings."""
+    cur_start = cur_end = 0
+    for chunk_size, file_path in zip(chunk_sizes, embeddings_paths):
+        cur_end += chunk_size
+        if cur_end < start:
+            cur_start += chunk_size
+            continue
+        slice_start = max(0, start - cur_start)
+        slice_end = min(chunk_size, end - cur_start)
+        with get_matrix_reader(
+            file_format, get_filesystem_class("hdfs")(), file_path, embedding_column_name, id_columns
```

Comment:
why force hdfs file system? spark also works on other file systems

Reply:
Good question. At first I didn't force it, I used fsspec.core.url_to_fs(path)[0] to resolve the file system, but it returned `HadoopFileSystem` instead of `PyArrowHDFS` and a file system `HadoopFileSystem` failed to open the file.
You are right, forcing it is not a good idea, I'll dig into it to find a solution.